src/functional/iterator/transforms/global_tmps.py

```diff
@@ -67,78 +195,152 @@ def _extend_domain(domain: ir.FunCall, offset_provider, shifts):
+    closures = []
+    shifts: dict[str, list[tuple]] = dict()
+    domain = None
+    for closure in reversed(node.fencil.closures):
+        if closure.domain == AUTO_DOMAIN:
+            output_shifts = shifts.get(closure.output.id, [])
+            domain = extend(domain, output_shifts)
+            closure = ir.StencilClosure(
+                domain=domain, stencil=closure.stencil, output=closure.output, inputs=closure.inputs
+            )
+        else:
             domain = closure.domain
-            while todos:
-                output, call = todos.pop()
-                output_shifts: List[tuple] = shifts.get(output.id, [])
-                assert isinstance(domain, ir.FunCall)
-                domain = self._extend_domain(domain, offset_provider, output_shifts)
-                if output.id in {tmp.id for tmp in tmps}:
-                    assert output.id not in tmp_domains
-                    tmp_domains[output.id] = domain
-                closure = ir.StencilClosure(
-                    domain=domain,
-                    stencil=call.fun,
-                    output=output,
-                    inputs=[handle_arg(arg) for arg in call.args],
-                )
-                local_shifts: Dict[str, List[tuple]] = dict()
-                CollectShifts().visit(closure.stencil, shifts=local_shifts)
-                assert isinstance(closure.stencil, ir.Lambda)
-                input_map: dict[str, str] = {
-                    param.id: arg.id for param, arg in zip(closure.stencil.params, closure.inputs)
-                }
-                for k, v in local_shifts.items():
-                    shifts.setdefault(input_map[k], []).extend(v)
-                closures.append(closure)
-
-        assert {tmp.id for tmp in tmps} == set(tmp_domains.keys())
-        if register_tmp is not None:
-            for tmp, domain in tmp_domains.items():
-                register_tmp(tmp, domain)
-
-        return ir.FencilDefinition(
-            id=node.id,
-            function_definitions=node.function_definitions,
-            params=node.params + tmps,
+            shifts = dict()
+
+        closures.append(closure)
+
+        if closure.stencil == ir.SymRef(id="deref"):
+            continue
+
+        local_shifts, params = _collect_stencil_shifts(closure.stencil, return_params=True)
+        input_map = {param.id: inp.id for param, inp in zip(params, closure.inputs)}
+        for param, shift in local_shifts.items():
+            shifts.setdefault(input_map[param], []).extend(shift)
+
+    return FencilWithTemporaries(
+        fencil=ir.FencilDefinition(
+            id=node.fencil.id,
+            function_definitions=node.fencil.function_definitions,
+            params=node.fencil.params[:-1],
             closures=list(reversed(closures)),
-        )
+        ),
+        params=node.params,
+        tmps=node.tmps,
+    )
+
+
+def update_unstructured_domains(node: FencilWithTemporaries, offset_provider):
```

Comment:
there is no test for this, should we add one or does this need to change anyway in the near future?